In [1]:
import pandas as pd

## Data Import and Initial Exploration

### Datasets Used:
1. `MissingMigrants-Global-2019-12-31_correct.csv`: The first dataset.
2. `MM_14_21.csv`: The second dataset.
3. `Global Missing Migrants Dataset 3.csv`: The third dataset.

### Reason:
Multiple datasets were used to obtain a comprehensive view of missing migrants across different dimensions and time frames.

### Operations:
1. Import the datasets into DataFrames.
2. Display basic information like shape and column names.


In [2]:
file_path = 'data/MissingMigrants-Global-2019-12-31_correct.csv'
missing_migrants_df = pd.read_csv(file_path)

In [3]:
missing_migrants_df

Web ID Region of Incident      Reported Date  Reported Year  \
0      52673      Mediterranean  December 30, 2019           2019   
1      52666      Mediterranean  December 30, 2019           2019   
2      52663          East Asia  December 27, 2019           2019   
3      52662        Middle East  December 26, 2019           2019   
4      52661        Middle East  December 24, 2019           2019   
...      ...                ...                ...            ...   
5982   39207   US-Mexico Border   January 16, 2014           2014   
5983   39368             Europe   January 16, 2014           2014   
5984   39257   US-Mexico Border   January 14, 2014           2014   
5985   39256   US-Mexico Border   January 12, 2014           2014   
5986   39206   US-Mexico Border   January 06, 2014           2014   

     Reported Month  Number Dead  Minimum Estimated Number of Missing  \
0               Dec          1.0                                  NaN   
1               Dec          1.0                                  NaN   
2               Dec          5.0                                  NaN   
3               Dec          7.0                                  NaN   
4               Dec         12.0                                  NaN   
...             ...          ...                                  ...   
5982            Jan          1.0                                  NaN   
5983            Jan          1.0                                  0.0   
5984            Jan          1.0                                  NaN   
5985            Jan          1.0                                  NaN   
5986            Jan          1.0                                  NaN   

      Total Dead and Missing  Number of Survivors  Number of Females  \
0                          1                 11.0                NaN   
1                          1                  NaN                NaN   
2                          5                  NaN                NaN   
3                          7                 64.0                NaN   
4                         12                  NaN                NaN   
...                      ...                  ...                ...   
5982                       1                  NaN                NaN   
5983                       1                  2.0                NaN   
5984                       1                  NaN                NaN   
5985                       1                  NaN                NaN   
5986                       1                  NaN                NaN   

      Number of Males  Number of Children              Cause of Death  \
0                 NaN                 NaN                 Hypothermia   
1                 1.0                 NaN           Presumed drowning   
2                 3.0                 NaN                     Unknown   
3                 NaN                 NaN                    Drowning   
4                 NaN                 NaN                  Air strike   
...               ...                 ...                         ...   
5982              1.0                 NaN                        Shot   
5983              1.0                 NaN                Hyperthermia   
5984              NaN                 NaN  Unknown (skeletal remains)   
5985              NaN                 NaN  Unknown (skeletal remains)   
5986              1.0                 NaN  Unknown (skeletal remains)   

                                   Location Description  \
0        Unspecififed location off the coast of Algeria   
1                 Recoverd on Calamorcarro Beach, Ceuta   
2     Bodies found on boat near Sado Island, Niigata...   
3               Van lake near Adilcevaz, Bitlis, Turkey   
4                        Al-Raqw market in Saada, Yemen   
...                                                 ...   
5982                              near Douglas, Arizona   
5983                  Border between Russia and Estonia   
5984                       

#### The table below provides an overview of the columns present in the first dataset along with their respective descriptions:

Column | Description
---------|----------
Web ID:  | A unique identifier for each record.
Region of Incident:  | The region where the incident occurred.
Reported Date:  | The date the incident was reported.
Reported Year:  | The year the incident was reported.
Reported Month:  | The month the incident was reported.
Number Dead:  | The number of people who died.
Minimum Estimated Number of Missing:  | The minimum estimated number of missing people.
Total Dead and Missing:  | The total number of dead and missing people.
Number of Survivors:  | The number of survivors.
Number of Females:  | The number of females involved.
Number of Males:  | The number of males involved.
Number of Children:  | The number of children involved.
Cause of Death:  | The cause of death.
Location Description:  | Description of the location where the incident occurred.
Information Source:  | The source of the information.
Location Coordinates:  | Coordinates of the location where the incident occurred.
Migration Route:  | The migration route involved.
URL:  | URL of the source.
UNSD Geographical Grouping:  | Geographical grouping according to the United Nations.
Source Quality:  | A rating for the quality of the source.

In [4]:
file_path_2 = 'data/MM_14_21.csv'
missing_migrants_df_2 = pd.read_csv(file_path_2)

In [5]:
missing_migrants_df_2

Unnamed: 0        Main ID    Incident ID         Region  \
0              0  2014.MMP00001  2014.MMP00001  North America   
1              1  2014.MMP00002  2014.MMP00002  North America   
2              2  2014.MMP00003  2014.MMP00003  North America   
3              3  2014.MMP00004  2014.MMP00004  North America   
4              4  2014.MMP00005  2014.MMP00005         Europe   
...          ...            ...            ...            ...   
9901        1265   2021.MMP0326   2021.MMP0326         Europe   
9902        1266   2021.MMP0327   2021.MMP0327         Europe   
9903        1267   2021.MMP0328   2021.MMP0328   Western Asia   
9904        1268   2021.MMP0329   2021.MMP0329   Western Asia   
9905        1269   2021.MMP0330   2021.MMP0330   Western Asia   

                Incident Date  Year Reported Month  Number Dead  \
0     Mon, 01/06/2014 - 12:00  2014        January          1.0   
1     Sun, 01/12/2014 - 12:00  2014        January          1.0   
2     Tue, 01/14/2014 - 12:00  2014        January          1.0   
3     Thu, 01/16/2014 - 12:00  2014        January          1.0   
4     Thu, 01/16/2014 - 12:00  2014        January          1.0   
...                       ...   ...            ...          ...   
9901  Sun, 09/19/2021 - 12:00  2021      September          1.0   
9902  Sun, 09/19/2021 - 12:00  2021      September          1.0   
9903  Wed, 09/08/2021 - 12:00  2021      September          NaN   
9904  Sat, 11/20/2021 - 12:00  2021       November          NaN   
9905  Fri, 12/31/2021 - 12:00  2021       December          NaN   

      Minimum Estimated Number of Missing  Total Number of Dead and Missing  \
0                                     NaN                                 1   
1                                     NaN                                 1   
2                                     NaN                                 1   
3                                     NaN                                 1   
4                                     0.0                                 1   
...                                   ...                               ...   
9901                                  NaN                                 1   
9902                                  NaN                                 1   
9903                                  1.0                                 1   
9904                                  1.0                                 1   
9905                                  1.0                                 1   

      ...  Number of Males  Number of Children  \
0     ...              1.0                 NaN   
1     ...              NaN                 NaN   
2     ...              NaN                 NaN   
3     ...              1.0                 NaN   
4     ...              1.0                 NaN   
...   ...              ...                 ...   
9901  ...              1.0                 NaN   
9902  ...              NaN                 NaN   
9903  ...              1.0                 NaN   
9904  ...              1.0                 NaN   
9905  ...              1.0                 1.0   

                                         Cause of Death  \
0                                      Mixed or unknown   
1                                      Mixed or unknown   
2                                      Mixed or unknown   
3                                              Violence   
4     Harsh environmental conditions / lack of adequ...   
...                                                 ...   
9901                                   Mixed or unknown   
9902  Harsh environmental conditions / lack of adequ...   
9903                                           Violence   
9904                                           Drowning   
9905  Harsh environmental conditions / lack of adequ...   

                                      Location of death  \
0     Pima Country Office of the Medical Examiner ju...   
1     Pima Country Office of the Medical Ex

#### This section describes the columns found in the second dataset:

Column: | Description
--------|------------
Unnamed: 0: | Seems like an index column.
Main ID: | A unique identifier for each record.
Incident ID: | Another identifier for each incident.
Region: | The region where the incident occurred.
Incident Date: | The date of the incident.
Year: | The year the incident occurred.
Reported Month: | The month the incident was reported.
Number Dead: | The number of people who died.
Minimum Estimated Number of Missing: | The minimum estimated number of missing people.
Total Number of Dead and Missing: | The total number of dead and missing people.
Number of Survivors: | The number of survivors.
Number of Females: | The number of females involved.
Number of Males: | The number of males involved.
Number of Children: | The number of children involved.
Cause of Death: | The cause of death.
Location of death: | Location where the death occurred.
Information Source: | The source of the information.
Coordinates: | Coordinates of the location where the incident occurred.
Migrantion route: | The migration route involved.
URL: | URL of the source.
UNSD Geographical Grouping: | Geographical grouping according to the United Nations.
Source Quality: | A rating for the quality of the source.

In [6]:
file_path_3 = 'data/Global Missing Migrants Dataset 3.csv'
missing_migrants_df_3 = pd.read_csv(file_path_3)

In [7]:
missing_migrants_df_3

Incident Type  Incident year Reported Month  \
0          Incident           2014        January   
1          Incident           2014        January   
2          Incident           2014        January   
3          Incident           2014        January   
4          Incident           2014        January   
...             ...            ...            ...   
13015      Incident           2023           July   
13016      Incident           2023           July   
13017      Incident           2023           July   
13018      Incident           2023           July   
13019      Incident           2023           July   

                    Region of Origin Region of Incident     Country of Origin  \
0                    Central America      North America             Guatemala   
1      Latin America / Caribbean (P)      North America               Unknown   
2      Latin America / Caribbean (P)      North America               Unknown   
3                    Central America      North America                Mexico   
4                    Northern Africa             Europe                 Sudan   
...                              ...                ...                   ...   
13015                   Western Asia       Western Asia  Syrian Arab Republic   
13016             Western Africa (P)       Western Asia               Unknown   
13017                 Western Africa    Northern Africa               Senegal   
13018                          Mixed    Northern Africa               Unknown   
13019             Western Africa (P)     Western Africa               Unknown   

       Number of Dead  Minimum Estimated Number of Missing  \
0                 1.0                                    0   
1                 1.0                                    0   
2                 1.0                                    0   
3                 1.0                                    0   
4                 1.0                                    0   
...               ...                                  ...   
13015             4.0                                    0   
13016             2.0                                    0   
13017            13.0                                    0   
13018             6.0                                    0   
13019            16.0                                   37   

       Total Number of Dead and Missing  Number of Survivors  \
0                                     1                    0   
1                                     1                    0   
2                                     1                    0   
3                                     1                    0   
4                                     1                    2   
...                                 ...                  ...   
13015                                 4                    0   
13016                                 2                    0   
13017                                13                    6   
13018                                 6                   48   
13019                                53                    2   

       Number of Females  Number of Males  Number of Children  \
0                      0                1                   0   
1                      0                0                   0   
2                      0                0                   0   
3                      0                1                   0   
4                      0                1                   0   
...                  ...              ...                 ...   
13015                  0                4                   0   
13016                  0                2                   0   
13017                  0                0                   0   
13018                  0                0                   0   
13019                  2                0                   0   

                                          Cause of Death  \
0                                       M

#### Below are the columns and their descriptions for the third dataset:

Columns in the Third Dataset | Description
-----------------------------|------------
Incident Type: | Type of incident.
Incident Year: | The year the incident occurred.
Reported Month: | The month the incident was reported.
Region of Origin: | The region of origin for the migrants.
Region of Incident: | The region where the incident occurred.
Country of Origin: | The country of origin for the migrants.
Number of Dead: | The number of people who died.
Minimum Estimated Number of Missing: | The minimum estimated number of missing people.
Total Number of Dead and Missing: | The total number of dead and missing people.
Number of Survivors: | The number of survivors.
Number of Females: | The number of females involved.
Number of Males: | The number of males involved.
Number of Children: | The number of children involved.
Cause of Death: | The cause of death.
Migration Route: | The migration route involved.
Location of Death: | Description of the location where the death occurred.
Information Source: | The source of the information.
Coordinates: | Coordinates of the location where the incident occurred.
UNSD Geographical Grouping: | Geographical grouping according to the United Nations.

### Observations:
Each dataset has a different structure, number of columns, and potentially a different/overlapping time range. Therefore, aligning them becomes important.

### Mapping Similar Columns Across Datasets

**Reason**: Since the datasets have different column names for similar information, there was a need to map them to a common set of columns for easy comparison and merging.

**Operations**:
- Define a dictionary to map similar columns across datasets.
- Rename the columns based on this mapping.



| First Dataset             | Second Dataset                 | Third Dataset                  | Description                                                   |
|---------------------------|--------------------------------|--------------------------------|---------------------------------------------------------------|
| Web ID                    | Main ID                         | -                              | Unique identifier for each record                             |
| Region of Incident        | Region                          | Region of Incident             | Geographical region where the incident occurred               |
| Reported Year             | Year                            | Incident year                  | Year the incident occurred                                    |
| Reported Date             | Incident Date                   | -                              | Date the incident occurred                                    |
| Reported Month            | Reported Month                  | Reported Month                 | Month the incident was reported                               |
| Number Dead               | Number Dead                     | Number of Dead                 | Number of people who died                                     |
| Minimum Estimated Number of Missing | Minimum Estimated Number of Missing | Minimum Estimated Number of Missing | Estimated number of missing people |
| Total Dead and Missing    | Total Number of Dead and Missing | Total Number of Dead and Missing | Total number of dead and missing people                       |
| Number of Survivors       | Number of Survivors              | Number of Survivors             | Number of survivors                                           |
| Number of Females         | Number of Females                | Number of Females               | Number of females involved                                    |
| Number of Males           | Number of Males                  | Number of Males                 | Number of males involved                                      |
| Number of Children        | Number of Children               | Number of Children              | Number of children involved                                   |
| Cause of Death            | Cause of Death                   | Cause of Death                  | Cause of the death(s)                                         |
| Location Description      | Location of death                | Location of death               | Description of the location where the incident occurred       |
| Information Source        | Information Source               | Information Source              | Source of the information                                     |
| Location Coordinates      | Coordinates                      | Coordinates                     | Coordinates of the location where the incident occurred      |
| Migration Route           | Migrantion route                 | Migration route                 | The migration route involved                                  |
| URL                       | URL                              | -                              | URL of the source                                             |
| UNSD Geographical Grouping | UNSD Geographical Grouping       | UNSD Geographical Grouping      | United Nations geographical grouping                           |
| Source Quality            | Source Quality                   | -                              | Quality of the source information                             |


In [8]:
missing_migrants_df.shape, missing_migrants_df_2.shape, missing_migrants_df_3.shape

((5987, 20), (9906, 22), (13020, 19))

In [9]:
def create_unique_columns_dataframe(datasets):
    unique_columns = {}
    
    for idx, dataset in enumerate(datasets, start=1):
        for column in dataset.columns:
            if column not in unique_columns:
                unique_columns[column] = [0] * len(datasets)
            unique_columns[column][idx-1] = 1
    
    unique_columns_df = pd.DataFrame.from_dict(unique_columns, orient='index', columns=[f"Dataset {idx}" for idx in range(1, len(datasets)+1)])
    return unique_columns_df.sort_index(ascending=True)


create_unique_columns_dataframe([missing_migrants_df, missing_migrants_df_2, missing_migrants_df_3])

Dataset 1  Dataset 2  Dataset 3
Cause of Death                               1          1          1
Coordinates                                  0          1          1
Country of Origin                            0          0          1
Incident Date                                0          1          0
Incident ID                                  0          1          0
Incident Type                                0          0          1
Incident year                                0          0          1
Information Source                           1          1          1
Location Coordinates                         1          0          0
Location Description                         1          0          0
Location of death                            0          1          1
Main ID                                      0          1          0
Migrantion route                             0          1          0
Migration Route                              1          0          0
Migration route                              0          0          1
Minimum Estimated Number of Missing          1          1          1
Number Dead                                  1          1          0
Number of Children                           1          1          1
Number of Dead                               0          0          1
Number of Females                            1          1          1
Number of Males                              1          1          1
Number of Survivors                          1          1          1
Region                                       0          1          0
Region of Incident                           1          0          1
Region of Origin                             0          0          1
Reported Date                                1          0          0
Reported Month                               1          1          1
Reported Year                                1          0          0
Source Quality                               1          1          0
Total Dead and Missing                       1          0          0
Total Number of Dead and Missing             0          1          1
UNSD Geographical Grouping                   1          1          1
URL                                          1          1          0
Unnamed: 0                                   0          1          0
Web ID                                       1          0          0
Year                                         0          1          0

In [10]:
column_mapping = {
    'Location Coordinates': 'Coordinates',
    'Incident year': 'Year',
    'Reported Year': 'Year',
    'Information Source': 'Source',
    'Location Description': 'Location of death',
    'Migrantion route': 'Migration route',
    'Migration Route': 'Migration route',
    'Number of Dead': 'Number Dead',
    'Region': 'Region of Incident',
    'Total Number of Dead and Missing': 'Total Dead and Missing'
}

In [11]:
def rename_columns(dataset, column_mapping):
    renamed_dataset = dataset.rename(columns=column_mapping)
    return renamed_dataset

In [12]:
missing_migrants_df = rename_columns(missing_migrants_df, column_mapping)
missing_migrants_df_2 = rename_columns(missing_migrants_df_2, column_mapping)
missing_migrants_df_3 = rename_columns(missing_migrants_df_3, column_mapping)

create_unique_columns_dataframe([missing_migrants_df, missing_migrants_df_2, missing_migrants_df_3])

Dataset 1  Dataset 2  Dataset 3
Cause of Death                               1          1          1
Coordinates                                  1          1          1
Country of Origin                            0          0          1
Incident Date                                0          1          0
Incident ID                                  0          1          0
Incident Type                                0          0          1
Location of death                            1          1          1
Main ID                                      0          1          0
Migration route                              1          1          1
Minimum Estimated Number of Missing          1          1          1
Number Dead                                  1          1          1
Number of Children                           1          1          1
Number of Females                            1          1          1
Number of Males                              1          1          1
Number of Survivors                          1          1          1
Region of Incident                           1          1          1
Region of Origin                             0          0          1
Reported Date                                1          0          0
Reported Month                               1          1          1
Source                                       1          1          1
Source Quality                               1          1          0
Total Dead and Missing                       1          1          1
UNSD Geographical Grouping                   1          1          1
URL                                          1          1          0
Unnamed: 0                                   0          1          0
Web ID                                       1          0          0
Year                                         1          1          1

#### Observation:

The datasets are now aligned in terms of column names, making it easier to perform subsequent analyses.


In [13]:
def count_values_across_dataframes(column_name, dataframes):
    unique_values = []
    
    # Collect unique values from all dataframes
    for df in dataframes:
        unique_values.extend(df[column_name].unique())
    
    print("Unique Values:", unique_values)  # Print the unique values
    
    unique_values = list(set(unique_values))  # Deduplicate the values
    unique_values = sorted(unique_values, key=str)  # Sort the values as strings
    
    result_df = pd.DataFrame(index=unique_values)
    
    # Fill in the dataframe with counts of the values in each dataframe
    for idx, df in enumerate(dataframes, start=1):
        value_counts = df[column_name].value_counts()
        result_df[f"Dataset {idx}"] = result_df.index.map(value_counts).fillna(0)
    
    # Convert the columns to integer data type
    result_df = result_df.astype(int)
    
    return result_df


## Extracting Time Range and Geographical Regions

### Reason
Knowing the time frame and geographical scope of each dataset is critical for understanding the context and limitations of the data.

### Operations
1. Extract the minimum and maximum years from each dataset.
2. Identify unique geographical regions mentioned.

In [14]:
time_range_1 = missing_migrants_df['Year'].min(), missing_migrants_df['Year'].max()
time_range_2 = missing_migrants_df_2['Year'].min(), missing_migrants_df_2['Year'].max()
time_range_3 = missing_migrants_df_3['Year'].min(), missing_migrants_df_3['Year'].max()

time_range_1, time_range_2, time_range_3

((2014, 2019), (2014, 2021), (2014, 2023))

In [15]:
(
    count_values_across_dataframes(
    'Region of Incident', 
    [missing_migrants_df, 
     missing_migrants_df_2, 
     missing_migrants_df_3])
)

Unique Values: ['Mediterranean', 'East Asia', 'Middle East', 'Europe', 'US-Mexico Border', 'Central America', 'Sub-Saharan Africa', 'Caribbean', 'North Africa', 'South Asia', 'Horn of Africa', 'South America', 'Southeast Asia', 'North America', 'Central Asia', 'North America', 'Europe', 'Mediterranean', 'Caribbean', 'South-eastern Asia', 'Southern Asia', 'Eastern Africa', 'Central America', 'Northern Africa', 'Eastern Asia', 'Western Asia', 'Western Africa', 'South America', 'Southern Africa', 'Middle Africa', 'Central Asia', 'North America', 'Europe', 'Mediterranean', 'Caribbean', 'South-eastern Asia', 'Southern Asia', 'Eastern Africa', 'Central America', 'Northern Africa', 'Eastern Asia', 'Western Asia', 'Western Africa', 'South America', 'Southern Africa', 'Middle Africa', 'Central Asia']


Dataset 1  Dataset 2  Dataset 3
Caribbean                  68        111        160
Central America           364        980       1375
Central Asia                1          1          1
East Asia                   6          0          0
Eastern Africa              0        397        489
Eastern Asia                0          8         10
Europe                    301        470        619
Horn of Africa            238          0          0
Mediterranean            1085       1759       2055
Middle Africa               0         72         75
Middle East               182          0          0
North Africa             1262          0          0
North America               2       2165       2706
Northern Africa             0       1810       2014
South America              36        165        209
South Asia                183          0          0
South-eastern Asia          0        205        237
Southeast Asia            137          0          0
Southern Africa             0          8         16
Southern Asia               0        511       1673
Sub-Saharan Africa        510          0          0
US-Mexico Border         1612          0          0
Western Africa              0        926        967
Western Asia                0        318        414

- Northern Africa and North Africa: These likely refer to the same geographic area.

- South-eastern Asia and Southeast Asia: Again, these likely refer to the same region.

- South Asia and Southern Asia: These also likely refer to the same region.

- East Asia and Eastern Asia: These also likely refer to the same region.

- Eastern Africa and Horn of Africa: While not exactly the same, these regions are closely related. The Horn of Africa is a peninsula in Eastern Africa and includes countries like Somalia, Ethiopia, Eritrea, and Djibouti. We might consider merging them depending on the level of granularity we need.

- Middle East and Western Asia: The Middle East often includes countries that are geographically in Western Asia, such as Saudi Arabia, Israel, and Iran.

- Sub-Saharan Africa and Middle Africa/Southern Africa/Western Africa/Eastern Africa: Sub-Saharan Africa is a broad term that includes various regions within Africa that are situated south of the Sahara. However, this would be a more significant consolidation and would depend on the granularity of analysis we will be looking for.

## Data Cleaning and Transformation
Reason: Data quality is essential for accurate analysis. This involves handling missing values, standardizing text, and converting data types where necessary.

### Operations
1. Check for inconsistencies in geographical naming and standardized them.

In [16]:
def update_column_values(column_name, value_mapping, dataframes):
    for df in dataframes:
        df[column_name] = df[column_name].map(lambda x: value_mapping.get(x, x))

In [17]:
region_mapping = {
    'Northern Africa': 'North Africa',
    'South-eastern Asia': 'Southeast Asia',
    'Southern Asia': 'South Asia',
    'East Asia': 'Eastern Asia'
}

(
    update_column_values(
        'Region of Incident', 
        region_mapping, 
        [missing_migrants_df, missing_migrants_df_2, missing_migrants_df_3]
    )
)

(
    count_values_across_dataframes(
    'Region of Incident', 
    [missing_migrants_df, 
     missing_migrants_df_2, 
     missing_migrants_df_3])
)

Unique Values: ['Mediterranean', 'Eastern Asia', 'Middle East', 'Europe', 'US-Mexico Border', 'Central America', 'Sub-Saharan Africa', 'Caribbean', 'North Africa', 'South Asia', 'Horn of Africa', 'South America', 'Southeast Asia', 'North America', 'Central Asia', 'North America', 'Europe', 'Mediterranean', 'Caribbean', 'Southeast Asia', 'South Asia', 'Eastern Africa', 'Central America', 'North Africa', 'Eastern Asia', 'Western Asia', 'Western Africa', 'South America', 'Southern Africa', 'Middle Africa', 'Central Asia', 'North America', 'Europe', 'Mediterranean', 'Caribbean', 'Southeast Asia', 'South Asia', 'Eastern Africa', 'Central America', 'North Africa', 'Eastern Asia', 'Western Asia', 'Western Africa', 'South America', 'Southern Africa', 'Middle Africa', 'Central Asia']


Dataset 1  Dataset 2  Dataset 3
Caribbean                  68        111        160
Central America           364        980       1375
Central Asia                1          1          1
Eastern Africa              0        397        489
Eastern Asia                6          8         10
Europe                    301        470        619
Horn of Africa            238          0          0
Mediterranean            1085       1759       2055
Middle Africa               0         72         75
Middle East               182          0          0
North Africa             1262       1810       2014
North America               2       2165       2706
South America              36        165        209
South Asia                183        511       1673
Southeast Asia            137        205        237
Southern Africa             0          8         16
Sub-Saharan Africa        510          0          0
US-Mexico Border         1612          0          0
Western Africa              0        926        967
Western Asia                0        318        414

In [18]:
(
    count_values_across_dataframes(
    'UNSD Geographical Grouping', 
    [missing_migrants_df, 
     missing_migrants_df_2, 
     missing_migrants_df_3])
)

Unique Values: ['Uncategorized', 'Eastern Asia', 'Western Asia', 'Southern Europe', 'Northern America', 'Western Europe', 'Central America', 'Western Africa', 'Caribbean', 'Northern Africa', 'Southern Asia', 'Eastern Africa', 'South America', 'South-eastern Asia', nan, 'Eastern Europe', 'Northern Europe', 'South Asia', 'Central Asia', 'Central Africa', 'Southern Africa', 'Northern America', 'Northern Europe', 'Uncategorized', 'Western Europe', 'Caribbean', 'South-eastern Asia', 'Southern Asia', 'Central America', 'Eastern Africa', 'Northern Africa', 'Eastern Asia', 'Western Asia', 'Eastern Europe', 'Southern Europe', 'Western Africa', 'South America', 'Southern Africa', 'Middle Africa', 'Central Asia', nan, 'Northern America', 'Northern Europe', 'Uncategorized', 'Western Europe', 'Caribbean', 'South-eastern Asia', 'Southern Asia', 'Central America', 'Eastern Africa', 'Northern Africa', 'Eastern Asia', 'Western Asia', 'Eastern Europe', 'Southern Europe', 'Western Africa', 'South America

Dataset 1  Dataset 2  Dataset 3
Caribbean                  58         79        113
Central Africa             15          0          0
Central America           640        974       1362
Central Asia                2          1          1
Eastern Africa            298        377        467
Eastern Asia                7          8         15
Eastern Europe             22         49        111
Middle Africa               0         72         75
Northern Africa          1262       1681       1872
Northern America         1310       2164       2708
Northern Europe            12         21         25
South America              36        160        207
South Asia                  9          0          0
South-eastern Asia         96        196        225
Southern Africa             6          8         12
Southern Asia             165        497       1660
Southern Europe           157        254        329
Uncategorized            1151       2030       2351
Western Africa            388        908        941
Western Asia              196        285        389
Western Europe            111        141        156
NaN                         0          0          0

## Comparison of Three Missing Migrants Datasets

### Time Range Covered

1. **First Dataset (`MissingMigrants-Global-2019-12-31_correct.csv`)**: 2014 - 2019
2. **Second Dataset (`MM_14_21.csv`)**: 2014 - 2021
3. **Third Dataset (`Global Missing Migrants Dataset 3.csv`)**: 2014 - 2023

### Geographical Coverage

- All three datasets cover a wide range of regions globally.
- The first dataset uses some unique names for regions (e.g., "Horn of Africa", "Sub-Saharan Africa") that are not present in the other two datasets.
- The second and third datasets are very similar in terms of the regions they cover.

### Data Completeness

#### First Dataset (`MissingMigrants-Global-2019-12-31_correct.csv`)

- Region of Incident: 0 missing values
- Reported Year: 0 missing values
- Number Dead: 225 missing values
- Total Dead and Missing: 0 missing values
- Cause of Death: 0 missing values

#### Second Dataset (`MM_14_21.csv`)

- Region: 0 missing values
- Year: 0 missing values
- Number Dead: 552 missing values
- Total Number of Dead and Missing: 0 missing values
- Cause of Death: 0 missing values

#### Third Dataset (`Global Missing Migrants Dataset 3.csv`)

- Region of Incident: 0 missing values
- Incident Year: 0 missing values
- Number of Dead: 550 missing values
- Total Number of Dead and Missing: 0 missing values
- Cause of Death: 0 missing values

### Statistical Summary

1. **Region**: All datasets cover multiple regions, with North America being a common region of focus.
  
2. **Year**: The years covered align with our earlier observations about the time range of each dataset.
  
3. **Number Dead**: The average number of dead varies between datasets but generally remains within a similar range. All datasets have incidents with missing values in this column.
  
4. **Total Dead and Missing**: The average number of total dead and missing also varies between datasets but is in a similar range.
  
5. **Cause of Death**: "Drowning" and "Mixed or unknown" are common causes of death across the datasets.



## Finding duplicate incidents in 3 datasets

### Reason: 
Duplicates can distort analytical results, so they need to be identified and handled appropriately.

### Operations:
- Identify duplicates based on a subset of columns.
- Drop duplicates to retain unique records.

In [19]:
def split_coordinates_and_convert(df, coordinate_column, lat_column, lon_column, rounded_lat_column, rounded_lon_column):
    df[[lat_column, lon_column]] = df[coordinate_column].str.split(',', expand=True)
    df[lat_column] = pd.to_numeric(df[lat_column], errors='coerce')
    df[lon_column] = pd.to_numeric(df[lon_column], errors='coerce')
    df[rounded_lat_column] = df[lat_column].round(3)
    df[rounded_lon_column] = df[lon_column].round(3)

# Applying the function to each dataset
split_coordinates_and_convert(missing_migrants_df, 'Coordinates', 'Lat', 'Lon', 'Rounded_Lat', 'Rounded_Lon')
split_coordinates_and_convert(missing_migrants_df_2, 'Coordinates', 'Lat', 'Lon', 'Rounded_Lat', 'Rounded_Lon')
split_coordinates_and_convert(missing_migrants_df_3, 'Coordinates', 'Lat', 'Lon', 'Rounded_Lat', 'Rounded_Lon')

# Display the first few rows to verify the correct splitting, conversion, and rounding
missing_migrants_df[['Lat', 'Lon', 'Rounded_Lat', 'Rounded_Lon']].head(), missing_migrants_df_2[['Lat', 'Lon', 'Rounded_Lat', 'Rounded_Lon']].head(), missing_migrants_df_3[['Lat', 'Lon', 'Rounded_Lat', 'Rounded_Lon']].head()


(         Lat         Lon  Rounded_Lat  Rounded_Lon
 0  35.568972   -1.289773       35.569       -1.290
 1  35.912384   -5.357673       35.912       -5.358
 2  38.154018  138.086033       38.154      138.086
 3  38.777229   42.739258       38.777       42.739
 4  17.245365   43.239093       17.245       43.239,
          Lat         Lon  Rounded_Lat  Rounded_Lon
 0  31.650259 -110.366453       31.650     -110.366
 1  31.597130 -111.737560       31.597     -111.738
 2  31.940260 -113.011250       31.940     -113.011
 3  31.506777 -109.315632       31.507     -109.316
 4  59.155100   28.000000       59.155       28.000,
          Lat         Lon  Rounded_Lat  Rounded_Lon
 0  31.650259 -110.366453       31.650     -110.366
 1  31.597130 -111.737560       31.597     -111.738
 2  31.940260 -113.011250       31.940     -113.011
 3  31.506777 -109.315632       31.507     -109.316
 4  59.155100   28.000000       59.155       28.000)

In [20]:
missing_migrants_df_3[missing_migrants_df_3['Year']<=2019]['Reported Month'].value_counts()

Reported Month
October      739
September    691
June         677
July         661
December     654
August       619
November     594
May          546
January      537
March        505
April        480
February     473
Name: count, dtype: int64

In [21]:
missing_migrants_df['Reported Month'].value_counts()

Reported Month
Jun    607
Oct    588
Sep    577
Aug    564
Jul    558
Dec    491
May    481
Jan    470
Nov    435
Apr    416
Feb    404
Mar    396
Name: count, dtype: int64

In [22]:
missing_migrants_df_2[missing_migrants_df_2['Year']<=2019]['Reported Month'].value_counts()

Reported Month
October      745
June         701
September    697
July         680
December     651
August       636
November     598
May          555
January      545
March        519
April        490
February     482
Name: count, dtype: int64

In [23]:
missing_migrants_df_3[missing_migrants_df_3['Year']<=2019]['Year'].value_counts()

Year
2019    1804
2018    1637
2017    1341
2016    1277
2015     821
2014     296
Name: count, dtype: int64

In [24]:
missing_migrants_df['Year'].value_counts()

Year
2018    1506
2017    1356
2016    1275
2015     808
2019     797
2014     245
Name: count, dtype: int64

In [25]:
missing_migrants_df_3[missing_migrants_df_3['Year']<=2019]['Region of Incident'].value_counts()

Region of Incident
North Africa       1604
North America      1436
Mediterranean      1094
Western Africa      890
Central America     659
Eastern Africa      351
Europe              326
Western Asia        229
South Asia          223
Southeast Asia      183
Caribbean            67
Middle Africa        63
South America        38
Eastern Asia          6
Southern Africa       6
Central Asia          1
Name: count, dtype: int64

In [26]:
missing_migrants_df['Region of Incident'].value_counts()

Region of Incident
US-Mexico Border      1612
North Africa          1262
Mediterranean         1085
Sub-Saharan Africa     510
Central America        364
Europe                 301
Horn of Africa         238
South Asia             183
Middle East            182
Southeast Asia         137
Caribbean               68
South America           36
Eastern Asia             6
North America            2
Central Asia             1
Name: count, dtype: int64

In [27]:
missing_migrants_df_3[missing_migrants_df_3['Year']<=2019]

Incident Type  Year Reported Month               Region of Origin  \
0         Incident  2014        January                Central America   
1         Incident  2014        January  Latin America / Caribbean (P)   
2         Incident  2014        January  Latin America / Caribbean (P)   
3         Incident  2014        January                Central America   
4         Incident  2014        January                Northern Africa   
...            ...   ...            ...                            ...   
7171      Incident  2019       November              Southern Asia (P)   
7172      Incident  2019      September                  Southern Asia   
7173      Incident  2019           July                  Southern Asia   
7174      Incident  2019           July              Southern Asia (P)   
7175      Incident  2019       December              Southern Asia (P)   

     Region of Incident Country of Origin  Number Dead  \
0         North America         Guatemala          1.0   
1         North America           Unknown          1.0   
2         North America           Unknown          1.0   
3         North America            Mexico          1.0   
4                Europe             Sudan          1.0   
...                 ...               ...          ...   
7171             Europe       Afghanistan          NaN   
7172             Europe          Pakistan          1.0   
7173             Europe          Pakistan          1.0   
7174             Europe           Unknown          1.0   
7175             Europe           Unknown          1.0   

      Minimum Estimated Number of Missing  Total Dead and Missing  \
0                                       0                       1   
1                                       0                       1   
2                                       0                       1   
3                                       0                       1   
4                                       0                       1   
...                                   ...                     ...   
7171                                    4                       4   
7172                                    0                       1   
7173                                    0                       1   
7174                                    0                       1   
7175                                    0                       1   

      Number of Survivors  ...  \
0                       0  ...   
1                       0  ...   
2                       0  ...   
3                       0  ...   
4                       2  ...   
...                   ...  ...   
7171                   11  ...   
7172                    0  ...   
7173                    0  ...   
7174                    0  ...   
7175                    0  ...   

                                         Cause of Death  \
0                                      Mixed or unknown   
1                                      Mixed or unknown   
2                                      Mixed or unknown   
3                                              Violence   
4     Harsh environmental conditions / lack of adequ...   
...                                                 ...   
7171                                           Drowning   
7172   Sickness / lack of access to adequate healthcare   
7173                                           Drowning   
7174                                           Drowning   
7175                                   Mixed or unknown   

                Migration route  \
0     US-Mexico border crossing   
1     US-Mexico border crossing   
2     US-Mexico border crossing   
3     US-Mexico border crossing   
4                           NaN   
...                         ...   
7171            Western Balkans   
7172            Western Balkans   
7173            Western Balkans   
7174            Western Balkans   
7175            Western Balkans   

                                      Location of d

## Analysis

In the concluding phase of our analysis, we aim to make key comparisons between the datasets. Specifically, we delve into the distribution of incidents both by year and by region, aiming to draw meaningful insights.

### Distribution by Year

We initiate this comparative analysis by investigating the fluctuation in the number of incidents across different years within the first dataset. By doing so, we gain insights into temporal trends and pinpoint the years characterized by the highest incident counts.

### Distribution by Region in Dataset 3

Moving forward, our attention turns to scrutinizing the distribution of incidents based on the 'Region of Incident' attribute within the third dataset. Our focus remains on the years leading up to 2019. This segment of the analysis serves the purpose of juxtaposing the regional distribution with that of the first dataset, thereby enabling the identification of common regions affected.

### Distribution by Region in Dataset 1

Furthermore, we conduct an assessment of the frequency of occurrences associated with distinct 'Region of Incident' categories in the first dataset. This examination enhances our comprehension of whether the two datasets align in terms of geographical emphasis.

### Filtering Dataset 3

The concluding step involves filtering the third dataset to exclusively encompass records pertaining to the years up to 2019. This modification is implemented to ensure an equitable comparison with the first dataset, which similarly encompasses data up to that specific year.

Through this comparative analysis, we aim to assess whether the third dataset can serve as a comprehensive and reliable source alone, for future analyses.


In [28]:
reported_month_dtype_df1 = missing_migrants_df['Reported Month'].dtype
reported_month_dtype_df2 = missing_migrants_df_2['Reported Month'].dtype
reported_month_dtype_df3 = missing_migrants_df_3['Reported Month'].dtype

reported_month_dtype_df1, reported_month_dtype_df2, reported_month_dtype_df3


(dtype('O'), dtype('O'), dtype('O'))

In [29]:
from datetime import datetime

def convert_month_strings_to_numbers(df):
    # Dictionary mapping month representations to numeric values
    month_mapping = {
        'Jan': 1, 'Feb': 2, 'Mar': 3, 'Apr': 4,
        'May': 5, 'Jun': 6, 'Jul': 7, 'Aug': 8,
        'Sep': 9, 'Oct': 10, 'Nov': 11, 'Dec': 12,
        'January': 1, 'February': 2, 'March': 3, 'April': 4, 'May': 5, 'June': 6, 'July': 7,
       'August': 8, 'September': 9, 'October': 10, 'November':11, 'December': 12
    }
    
    # Convert 'Reported Month' column to numeric month representations
    df['Reported Month'] = df['Reported Month'].map(month_mapping)
    
    return df

## Merging Datasets
Reason: The aim was to create a comprehensive dataset by combining information from all three datasets.

### Operations
1. Identify records that are duplicates across datasets.
2. Merge the datasets while avoiding duplicate entries.

In [30]:
missing_migrants_df = convert_month_strings_to_numbers(missing_migrants_df)  # Replace df1 with your actual DataFrame name
missing_migrants_df_2 = convert_month_strings_to_numbers(missing_migrants_df_2)  # Replace df2 with your actual DataFrame name
missing_migrants_df_3 = convert_month_strings_to_numbers(missing_migrants_df_3)  # Replace df3 with your actual DataFrame name

In [31]:
duplicates_df1 = missing_migrants_df.duplicated()
duplicates_df2 = missing_migrants_df_2.duplicated()
duplicates_df3 = missing_migrants_df_3.duplicated()

print(f"Number of duplicate rows in DataFrame 1: {duplicates_df1.sum()}")
print(f"Number of duplicate rows in DataFrame 2: {duplicates_df2.sum()}")
print(f"Number of duplicate rows in DataFrame 3: {duplicates_df3.sum()}")

Number of duplicate rows in DataFrame 1: 0
Number of duplicate rows in DataFrame 2: 0
Number of duplicate rows in DataFrame 3: 641


In [32]:
missing_migrants_df_3 = missing_migrants_df_3.drop_duplicates()

# Check the shape of the DataFrame after dropping duplicates
print("Shape after dropping duplicates:", missing_migrants_df_3.shape)

Shape after dropping duplicates: (12379, 23)


In [33]:
subset = ['Year', 'Reported Month', 'Region of Incident', 'Total Dead and Missing',
                      'Number Dead', 'UNSD Geographical Grouping', 'Cause of Death',
                      'Number of Children', 'Number of Males', 'Number of Females', 'Lat', 'Lon']
duplicates_df1 = missing_migrants_df.duplicated(subset=subset)
duplicates_df2 = missing_migrants_df_2.duplicated(subset=subset)
duplicates_df3 = missing_migrants_df_3.duplicated(subset=subset)

print(f"Number of duplicate rows in DataFrame 1: {duplicates_df1.sum()}")
print(f"Number of duplicate rows in DataFrame 2: {duplicates_df2.sum()}")
print(f"Number of duplicate rows in DataFrame 3: {duplicates_df3.sum()}")

Number of duplicate rows in DataFrame 1: 469
Number of duplicate rows in DataFrame 2: 530
Number of duplicate rows in DataFrame 3: 124


In [34]:
print("Shape before dropping duplicates:", missing_migrants_df.shape)
missing_migrants_df.drop_duplicates(subset)
print("Shape after dropping duplicates:", missing_migrants_df.shape)

Shape before dropping duplicates: (5987, 24)
Shape after dropping duplicates: (5987, 24)


In [35]:
print("Shape before dropping duplicates:", missing_migrants_df_2.shape)
missing_migrants_df_2.drop_duplicates(subset)
print("Shape after dropping duplicates:", missing_migrants_df_2.shape)

Shape before dropping duplicates: (9906, 26)
Shape after dropping duplicates: (9906, 26)


In [36]:
print("Shape before dropping duplicates:", missing_migrants_df_3.shape)
missing_migrants_df_3.drop_duplicates(subset)
print("Shape after dropping duplicates:", missing_migrants_df_3.shape)

Shape before dropping duplicates: (12379, 23)
Shape after dropping duplicates: (12379, 23)


In [37]:
matched_df3_df1_ll = missing_migrants_df_3.merge(missing_migrants_df, on=subset, how='inner')
matched_df3_df1_ll


Incident Type  Year  Reported Month             Region of Origin  \
0         Incident  2014               7                      Unknown   
1   Split Incident  2015               9                 Western Asia   
2         Incident  2015               9                 Western Asia   
3         Incident  2015               9                Southern Asia   
4         Incident  2015              10                 Western Asia   
5         Incident  2015              10                 Western Asia   
6         Incident  2015              10  Western / Southern Asia (P)   
7         Incident  2015              10                 Western Asia   
8         Incident  2015              10  Western / Southern Asia (P)   
9         Incident  2015              10                Southern Asia   
10        Incident  2015              10  Western / Southern Asia (P)   
11        Incident  2015              11  Western / Southern Asia (P)   
12        Incident  2015              12  Western / Southern Asia (P)   
13        Incident  2015              12  Western / Southern Asia (P)   
14        Incident  2015              12  Western / Southern Asia (P)   
15        Incident  2015              12  Western / Southern Asia (P)   
16        Incident  2015              12  Western / Southern Asia (P)   
17        Incident  2016               1  Western / Southern Asia (P)   
18        Incident  2016               1                 Western Asia   
19        Incident  2016               1  Western / Southern Asia (P)   
20        Incident  2016               1  Western / Southern Asia (P)   
21        Incident  2016               7                      Unknown   
22  Split Incident  2016               7              Central America   
23        Incident  2016               9                        Mixed   
24        Incident  2016              10                        Mixed   
25        Incident  2016              10                      Unknown   
26        Incident  2016              10               Western Africa   
27        Incident  2016              11                Southern Asia   
28        Incident  2017               1                      Unknown   
29        Incident  2017               2                    Caribbean   
30        Incident  2017               4                        Mixed   
31        Incident  2017               4       Sub-Saharan Africa (P)   
32        Incident  2017               4                        Mixed   
33        Incident  2017               6                      Unknown   
34        Incident  2017               7                    Caribbean   
35        Incident  2019              10                    Caribbean   

   Region of Incident                            Country of Origin  \
0       Mediterranean                                      Unknown   
1       Mediterranean                    Iraq,Syrian Arab Republic   
2       Mediterranean                    Iraq,Syrian Arab Republic   
3       Mediterranean                                  Afghanistan   
4       Mediterranean                         Syrian Arab Republic   
5       Mediterranean                                      Lebanon   
6       Mediterranean                                      Unknown   
7       Mediterranean                         Syrian Arab Republic   
8       Mediterranean                                      Unknown   
9       Mediterranean                                  Afghanistan   
10      Mediterranean                                      Unknown   
11      Mediterranean                                      Unknown   
12      Mediterranean                                      Unknown   
13      Mediterranean                                      Unknown   
14      Mediterranean                                      Unknown   
15      Mediterranean                                      Unknown   
16      Mediterranean                                      Unknown   
17      Mediterranean                      

In [38]:
missing_migrants_df['dataset'] = 1
missing_migrants_df_2['dataset'] = 2
missing_migrants_df_3['dataset'] = 3

merged_df = pd.concat([missing_migrants_df_3, missing_migrants_df_2, missing_migrants_df])

# Check for duplicates
duplicates_mask = merged_df.duplicated(subset=subset, keep=False)

# Remove duplicates
unique_records_df = merged_df[~duplicates_mask]

# Display the unique records DataFrame
unique_records_df

Incident Type  Year  Reported Month               Region of Origin  \
0         Incident  2014               1                Central America   
1         Incident  2014               1  Latin America / Caribbean (P)   
2         Incident  2014               1  Latin America / Caribbean (P)   
3         Incident  2014               1                Central America   
4         Incident  2014               1                Northern Africa   
...            ...   ...             ...                            ...   
5972           NaN  2014               1                            NaN   
5980           NaN  2014               1                            NaN   
5981           NaN  2014               1                            NaN   
5982           NaN  2014               1                            NaN   
5983           NaN  2014               1                            NaN   

     Region of Incident Country of Origin  Number Dead  \
0         North America         Guatemala          1.0   
1         North America           Unknown          1.0   
2         North America           Unknown          1.0   
3         North America            Mexico          1.0   
4                Europe             Sudan          1.0   
...                 ...               ...          ...   
5972             Europe               NaN          1.0   
5980      Mediterranean               NaN         12.0   
5981   US-Mexico Border               NaN          1.0   
5982   US-Mexico Border               NaN          1.0   
5983             Europe               NaN          1.0   

      Minimum Estimated Number of Missing  Total Dead and Missing  \
0                                     0.0                       1   
1                                     0.0                       1   
2                                     0.0                       1   
3                                     0.0                       1   
4                                     0.0                       1   
...                                   ...                     ...   
5972                                  0.0                       1   
5980                                  NaN                      12   
5981                                  NaN                       1   
5982                                  NaN                       1   
5983                                  0.0                       1   

      Number of Survivors  ...  Rounded_Lon  dataset  Unnamed: 0 Main ID  \
0                     0.0  ...     -110.366        3         NaN     NaN   
1                     0.0  ...     -111.738        3         NaN     NaN   
2                     0.0  ...     -113.011        3         NaN     NaN   
3                     0.0  ...     -109.316        3         NaN     NaN   
4                     2.0  ...       28.000        3         NaN     NaN   
...                   ...  ...          ...      ...         ...     ...   
5972                  2.0  ...        1.879        1         NaN     NaN   
5980                  NaN  ...       27.158        1         NaN     NaN   
5981                  NaN  ...     -111.624        1         NaN     NaN   
5982                  NaN  ...     -109.545        1         NaN     NaN   
5983                  2.0  ...       27.859        1         NaN     NaN   

     Incident ID Incident Date                    URL Source Quality   Web ID  \
0            NaN           NaN                    NaN            NaN      NaN   
1            NaN           NaN                    NaN            NaN      NaN   
2            NaN           NaN                    NaN            NaN      NaN   
3            NaN           NaN                    NaN            NaN      NaN   
4            NaN           NaN                    NaN            NaN      NaN   
...          ...           ...                    ...            ...      ...   
5972         NaN           NaN  http://bit.ly/1icTIF9            4.0  39373.0   
5980         NaN 

In [39]:
subset_rounded = ['Year', 'Reported Month', 'Region of Incident', 'Total Dead and Missing',
                      'Number Dead', 'UNSD Geographical Grouping', 'Cause of Death',
                      'Number of Children', 'Number of Males', 'Number of Females', 'Reounded_Lat', 'Rounded_Lon']
print("Shape before dropping duplicates:", unique_records_df.shape)
unique_records_df = unique_records_df.drop_duplicates(subset)
print("Shape after dropping duplicates:", unique_records_df.shape)

Shape before dropping duplicates: (25184, 32)
Shape after dropping duplicates: (25184, 32)


In [40]:
# Group by 'Year', 'Reported Month', and 'Dataset' and calculate the count of records
grouped = merged_df.groupby(['Year', 'Reported Month', 'dataset']).size().reset_index(name='Record Count')

# Pivot the grouped data to get the desired format
pivot_table = grouped.pivot_table(index=['Reported Month', 'Year'], columns='dataset', values='Record Count', fill_value=0)

# Rename the columns
pivot_table.columns = [f'Dataset {col}' for col in pivot_table.columns]

# Reset the index to make the pivot table more readable
pivot_table.reset_index(inplace=True)

pivot_table

Reported Month  Year  Dataset 1  Dataset 2  Dataset 3
0                 1  2014         16         17         17
1                 1  2015         35         37         37
2                 1  2016        116        116        109
3                 1  2017        129        124        123
4                 1  2018        117        123        115
..              ...   ...        ...        ...        ...
110              12  2018         58         99         99
111              12  2019         27        141        137
112              12  2020          0         91         72
113              12  2021          0         89        140
114              12  2022          0          0        120

[115 rows x 5 columns]

In [41]:
jan_2015 = unique_records_df[(unique_records_df['Reported Month']==1) & (unique_records_df['Year']==2015)]

In [42]:
top_3_duplicates = jan_2015.sort_values(by='Number Dead', ascending=False).head(3)

In [43]:
top_3_duplicates.T

338   \
Incident Type                                                                      NaN   
Year                                                                              2015   
Reported Month                                                                       1   
Region of Origin                                                                   NaN   
Region of Incident                                                        North Africa   
Country of Origin                                                                  NaN   
Number Dead                                                                       26.0   
Minimum Estimated Number of Missing                                                NaN   
Total Dead and Missing                                                              26   
Number of Survivors                                                                NaN   
Number of Females                                                                  NaN   
Number of Males                                                                    NaN   
Number of Children                                                                 NaN   
Cause of Death                       Harsh environmental conditions / lack of adequ...   
Migration route                                                 Sahara Desert crossing   
Location of death                                          Unspecified location, Libya   
Source                                                                            MHub   
Coordinates                                                       26.239989, 12.913084   
UNSD Geographical Grouping                                             Northern Africa   
Lat                                                                          26.239989   
Lon                                                                          12.913084   
Rounded_Lat                                                                      26.24   
Rounded_Lon                                                                     12.913   
dataset                                                                              2   
Unnamed: 0                                                                        32.0   
Main ID                                                                  2015.MMP00035   
Incident ID                                                              2015.MMP00035   
Incident Date                                                  Sat, 01/31/2015 - 12:00   
URL                                  https://archive.vn/PXCXg, https://archive.vn/0...   
Source Quality                                                                     2.0   
Web ID                                                                             NaN   
Reported Date                                                                      NaN   

                                                                                  328   \
Incident Type                                                                 Incident   
Year                                                                              2015   
Reported Month                                                                       1   
Region of Origin                                                               Unknown   
Region of Incident                                                        North Africa   
Country of Origin                                                              Unknown   
Number Dead                                                                       26.0   
Minimum Estimated Number of Missing                                                0.0   
Total Dead and Missing                                                              26   
Number of Survivors                                                                0.0   
Number of Females                                                                  0.0   
Number of Males                

In [44]:
subset_rounded = ['Year', 'Reported Month', 'Region of Incident', 'Total Dead and Missing',
                      'Number Dead', 'UNSD Geographical Grouping', 'Rounded_Lat', 'Rounded_Lon']
print("Shape before dropping duplicates:", unique_records_df.shape)
unique_records_df = unique_records_df.drop_duplicates(subset_rounded)
print("Shape after dropping duplicates:", unique_records_df.shape)

Shape before dropping duplicates: (25184, 32)
Shape after dropping duplicates: (14194, 32)


In [45]:
duplicated_indices = unique_records_df.index.duplicated(keep=False)

# Use boolean indexing to extract the records with duplicated indices
duplicated_records = unique_records_df[duplicated_indices]
duplicated_records.sort_index(ascending=True)

Incident Type  Year  Reported Month Region of Origin Region of Incident  \
3         Incident  2014               1  Central America      North America   
3              NaN  2019              12              NaN        Middle East   
4              NaN  2019              12              NaN        Middle East   
4         Incident  2014               1  Northern Africa             Europe   
7              NaN  2019              12              NaN   US-Mexico Border   
...            ...   ...             ...              ...                ...   
9895      Incident  2021               6    Southern Asia         South Asia   
9898           NaN  2021              12              NaN     Southeast Asia   
9898      Incident  2021               6    Southern Asia         South Asia   
9903           NaN  2021               9              NaN       Western Asia   
9903      Incident  2021               6    Southern Asia         South Asia   

     Country of Origin  Number Dead  Minimum Estimated Number of Missing  \
3               Mexico          1.0                                  0.0   
3                  NaN          7.0                                  NaN   
4                  NaN         12.0                                  NaN   
4                Sudan          1.0                                  0.0   
7                  NaN         83.0                                  NaN   
...                ...          ...                                  ...   
9895       Afghanistan          1.0                                  0.0   
9898               NaN         10.0                                 12.0   
9898       Afghanistan          1.0                                  0.0   
9903               NaN          NaN                                  1.0   
9903       Afghanistan          1.0                                  0.0   

      Total Dead and Missing  Number of Survivors  ...  Rounded_Lon  dataset  \
3                          1                  0.0  ...     -109.316        3   
3                          7                 64.0  ...       42.739        1   
4                         12                  NaN  ...       43.239        1   
4                          1                  2.0  ...       28.000        3   
7                         83                  NaN  ...      -99.428        1   
...                      ...                  ...  ...          ...      ...   
9895                       1                  0.0  ...       51.101        3   
9898                      22                 35.0  ...      101.043        2   
9898                       1                  0.0  ...       57.571        3   
9903                       1                  NaN  ...       26.464        2   
9903                       1                  0.0  ...       61.163        3   

      Unnamed: 0       Main ID   Incident ID            Incident Date  \
3            NaN           NaN           NaN                      NaN   
3            NaN           NaN           NaN                      NaN   
4            NaN           NaN           NaN                      NaN   
4            NaN           NaN           NaN                      NaN   
7            NaN           NaN           NaN                      NaN   
...          ...           ...           ...                      ...   
9895         NaN           NaN           NaN                      NaN   
9898      1262.0  2021.MMP0323  2021.MMP0323  Sat, 12/25/2021 - 12:00   
9898         NaN           NaN           NaN                      NaN   
9903      1267.0  2021.MMP0328  2021.MMP0328  Wed, 09/08/2021 - 12:00   
9903         NaN           NaN           NaN                      NaN   

                                                    URL Source Quality  \
3                                                   NaN            NaN   
3     http://bit.ly/2ZG2Y19, http://bit.ly/2MLamDf, ...            3.0   
4     http://bit.ly/2FjolvD, http://bit.ly/2sD42GR, ...            4.0 

In [46]:
output_filename = 'data/merged_records.csv'
unique_records_df.to_csv(output_filename, index=False)
print(f"DataFrame has been written to {output_filename}")

DataFrame has been written to data/merged_records.csv


In [47]:
unique_records_df

Incident Type  Year  Reported Month               Region of Origin  \
0         Incident  2014               1                Central America   
1         Incident  2014               1  Latin America / Caribbean (P)   
2         Incident  2014               1  Latin America / Caribbean (P)   
3         Incident  2014               1                Central America   
4         Incident  2014               1                Northern Africa   
...            ...   ...             ...                            ...   
5972           NaN  2014               1                            NaN   
5980           NaN  2014               1                            NaN   
5981           NaN  2014               1                            NaN   
5982           NaN  2014               1                            NaN   
5983           NaN  2014               1                            NaN   

     Region of Incident Country of Origin  Number Dead  \
0         North America         Guatemala          1.0   
1         North America           Unknown          1.0   
2         North America           Unknown          1.0   
3         North America            Mexico          1.0   
4                Europe             Sudan          1.0   
...                 ...               ...          ...   
5972             Europe               NaN          1.0   
5980      Mediterranean               NaN         12.0   
5981   US-Mexico Border               NaN          1.0   
5982   US-Mexico Border               NaN          1.0   
5983             Europe               NaN          1.0   

      Minimum Estimated Number of Missing  Total Dead and Missing  \
0                                     0.0                       1   
1                                     0.0                       1   
2                                     0.0                       1   
3                                     0.0                       1   
4                                     0.0                       1   
...                                   ...                     ...   
5972                                  0.0                       1   
5980                                  NaN                      12   
5981                                  NaN                       1   
5982                                  NaN                       1   
5983                                  0.0                       1   

      Number of Survivors  ...  Rounded_Lon  dataset  Unnamed: 0 Main ID  \
0                     0.0  ...     -110.366        3         NaN     NaN   
1                     0.0  ...     -111.738        3         NaN     NaN   
2                     0.0  ...     -113.011        3         NaN     NaN   
3                     0.0  ...     -109.316        3         NaN     NaN   
4                     2.0  ...       28.000        3         NaN     NaN   
...                   ...  ...          ...      ...         ...     ...   
5972                  2.0  ...        1.879        1         NaN     NaN   
5980                  NaN  ...       27.158        1         NaN     NaN   
5981                  NaN  ...     -111.624        1         NaN     NaN   
5982                  NaN  ...     -109.545        1         NaN     NaN   
5983                  2.0  ...       27.859        1         NaN     NaN   

     Incident ID Incident Date                    URL Source Quality   Web ID  \
0            NaN           NaN                    NaN            NaN      NaN   
1            NaN           NaN                    NaN            NaN      NaN   
2            NaN           NaN                    NaN            NaN      NaN   
3            NaN           NaN                    NaN            NaN      NaN   
4            NaN           NaN                    NaN            NaN      NaN   
...          ...           ...                    ...            ...      ...   
5972         NaN           NaN  http://bit.ly/1icTIF9            4.0  39373.0   
5980         NaN 

In [48]:
unique_records_df['dataset'].value_counts()

dataset
3    11182
1     2546
2      466
Name: count, dtype: int64

## Key Takeaways
1. Comprehensive Source: Dataset 3 appears to be the most comprehensive, containing most of the records that are also in Datasets 1 and 2.

2. Redundancy: The lower number of unique records contributed by Datasets 1 and 2 suggests that they were mainly subsets of Dataset 3, or that there is a significant overlap between these datasets and Dataset 3.

3. The records from Datasets 1 and 2 extend Dataset 3

4. Data Integrity: This overlap could be a positive factor for data integrity, as it might imply a level of confirmation or validation for the records that appear in multiple datasets.
